lib/route/link/bridge.c

```diff
@@ -410,6 +577,39 @@ int _nl_bridge_fill_vlan_info(struct nl_msg *msg, struct rtnl_link_bridge_vlan *
 	return -NLE_MSGSIZE;
 }
 
+static int bridge_fill_mst(struct nl_msg *msg, struct nl_list_head *mst_list) {
+	struct nlattr *attr = NULL;
+	struct nlattr *entry_attr = NULL;
+	struct mst_state_entry *entry;
+
+	if (nl_list_empty(mst_list))
+		return 0;
+
+	attr = nla_nest_start(msg, IFLA_BRIDGE_MST);
+	if (!attr)
+		goto nla_put_failure;
+
+	nl_list_for_each_entry(entry, mst_list, list_node) {
+		entry_attr = nla_nest_start(msg, IFLA_BRIDGE_MST_ENTRY);
+		if (!entry_attr)
+			goto nla_put_failure;
+
+		NLA_PUT_U16(msg, IFLA_BRIDGE_MST_ENTRY_MSTI, entry->msti);
+		NLA_PUT_U8(msg, IFLA_BRIDGE_MST_ENTRY_STATE, entry->state);
+
+		nla_nest_end(msg, entry_attr);
+	}
+
+	nla_nest_end(msg, attr);
+
+	return 0;
+
+nla_put_failure:
+	nla_nest_cancel(msg, entry_attr);
+	nla_nest_cancel(msg, attr);
```

Comment:
there are two explicit `goto nla_put_failure`. Those should not both cancel the nested attributes...

Reply:
I'm not entirely sure what you mean, could you explain some more? The nla_nest_cancel calls at the bottom will only cancel attributes that are not null, so it should work. I could change the code to the following which might be more clear:
```c
static int bridge_fill_mst(struct nl_msg *msg, struct nl_list_head *mst_list) {
	struct nlattr *attr = NULL;
	struct nlattr *entry_attr = NULL;
	struct mst_state_entry *entry;

	if (nl_list_empty(mst_list))
		return 0;

	attr = nla_nest_start(msg, IFLA_BRIDGE_MST);
	if (!attr)
		goto err_out;

	nl_list_for_each_entry(entry, mst_list, list_node) {
		entry_attr = nla_nest_start(msg, IFLA_BRIDGE_MST_ENTRY);
		if (!entry_attr)
			goto err_out_nest_cancel_attr;

		NLA_PUT_U16(msg, IFLA_BRIDGE_MST_ENTRY_MSTI, entry->msti);
		NLA_PUT_U8(msg, IFLA_BRIDGE_MST_ENTRY_STATE, entry->state);

		nla_nest_end(msg, entry_attr);
	}

	nla_nest_end(msg, attr);

	return 0;

nla_put_failure:
	nla_nest_cancel(msg, entry_attr);
err_out_nest_cancel_attr:
	nla_nest_cancel(msg, attr);
err_out:
	return -NLE_MSGSIZE;
}
```

Reply:
I went ahead and made this change since I think it's clearer.

Reply:
ah, I didn't realize that `nla_nest_cancel()` handles a `NULL` argument. So it was correct previously.

I still like this better.